aten/src/ATen/native/cpu/SoftMaxKernel.cpp

```diff
@@ -207,6 +207,142 @@ struct vec_host_softmax_lastdim {
+
+              sum_m256_o1 = sum_m256_o1 + output_m256_o1;
+              sum_m256_o2 = sum_m256_o2 + output_m256_o2;
+            }
+            // Step3: Unify
+            for (int64_t d = 0; d < dim_size; d++) {
+              Vec output_m256_o1 = Vec::loadu(temp_vec_output_data + d*vectorized_step*2);
+              Vec output_m256_o2 = Vec::loadu(temp_vec_output_data + d*vectorized_step*2 + vectorized_step);
+              output_m256_o1 = output_m256_o1/sum_m256_o1;
+              output_m256_o2 = output_m256_o2/sum_m256_o2;
+              Vec_bf16 output_m256_bf16 = convert_float_bfloat16(output_m256_o1, output_m256_o2);
+              output_m256_bf16.store(output_data + d * dim_stride);
+            }
+            idx += vectorized_step;
+          } else {
+            // Tail case(Scalar): it is exactly same logic as host_softmax
+            // inside aten/src/ATen/native/SoftMax.cpp. There are 2 kind of
+            // cases which will fall through this part:
+            // Case 1: For the idx at the end of total chunk for each thread, there are not enough numbers for parallization.
+            // Case 2: For the idx at the end of each inner_size inside thread, there are not enough numbers for parallization.
+            int64_t tail_number = ((idx+vectorized_step) > end) ? /*Case1*/ (end - idx) : /*Case2*/ (inner_size - inner_idx);
+            for (int64_t i=0; i < tail_number; i++) {
+              outer_idx = (idx + i) / inner_size;
+              inner_idx = (idx + i) % inner_size;
+              BFloat16* input_data =
+                  input_data_base + outer_idx * outer_stride + inner_idx;
+              BFloat16* output_data =
+                  output_data_base + outer_idx * outer_stride + inner_idx;
+              // Step1: Get max score
+              float max_input = float(input_data[0]);
+              for (int64_t d = 1; d < dim_size; d++) {
+                max_input = std::max(max_input, float(input_data[d * dim_stride]));
+              }
+              // Step2: Calculate the Sum
+              float sum_data = 0.0;
+              float temp_output_data = 0.0;
+              for (int64_t d = 0; d < dim_size; d++) {
+                temp_output_data = std::exp(input_data[d * dim_stride] - max_input);
+                sum_data += temp_output_data;
+                output_data[d * dim_stride] = c10::BFloat16(temp_output_data);
+              }
+              // Step3: Unify
+              for (int64_t d = 0; d < dim_size; d++) {
+                output_data[d * dim_stride] =
+                    c10::BFloat16(float(output_data[d * dim_stride])/sum_data);
+              }
+            }
+            idx += tail_number;
+          }
+        }
+      });
+}
+
+template <typename scalar_t>
+inline void assert_vectorized_step(int vectorized_step){
+    TORCH_CHECK(false, "unsupported dtype to do the vectorized calculation");
```

Comment:
nit: If you bring the passed dtype as part of the error message it'll be easier to understand what went wrong, i.e. "Unsupported dtype", <appropriate_to_string_function>(scalar_t), "expected dtype to be one of double, float or BFloat16".

Reply:
At first thought, I want to use `typeid(scalar_t).name()`. But the message it prints are not guaranteed to be human-readable. Also I didn't find other `<appropriate_to_string_function>` function to do the scalar_t to string convert inside pytorch. 

Another thing is that `assert_vectorized_step` looks like redundant now. Since the `vectorized_step` is getting from ` vec::Vectorized<scalar_t>.size()`, it already make sure the steps are corresponding to specific dtype. So we can delete them to reduce code size. We meet some mistakes previously because of hard-coded `vectorized_step`.

Reply:
Also we already guarantee the inputs to be one of `float, double, bfloat16` outside. https://github.com/pytorch/pytorch/blob/2aebb58345f09b23160506611ba28fcfd053f2f4/aten/src/ATen/native/cpu/SoftMaxKernel.cpp#L480-L484

Reply:
If we have to do the `assert`, there are 2 ways come into my mind:
- Use `CppTypeToScalarType` to convert CppType back to pytorch scalar_type. Then we can print the message
- Write CppType to string template. 
But looks like we don't need these redundant assert.